Add MySQL storage and tests

lucasmrod opened the pull request:
Hi folks.

This PR implements a MySQL storage, issue #4.
I've added tests that can be run on any future storage (which caught a small bug in the `FileStorage`).

zwass commented on http/api/tokenpki.go:
```diff
@@ -55,14 +46,15 @@ func GetCertTokenPKIHandler(store TokenPKIStorer, logger log.Logger) http.Handle
 			return
 		}
 		logger = logger.With("name", r.URL.Path)
+		// TODO(lucas): defaultDays is 1.
```
Intending to do this?

lucasmrod replied:
Yes. I'll remove the comment and open a separate PR for this.

lucasmrod replied:
See https://github.com/micromdm/nanodep/pull/6.

zwass commented on storage/mysql/mysql.go:
```diff
@@ -0,0 +1,294 @@
+
+// StoreConfig saves the DEP config for name DEP name.
+func (s *MySQLStorage) StoreConfig(ctx context.Context, name string, config *client.Config) error {
+	if config == nil || config.BaseURL == "" {
+		return nil
+	}
```
I don't have much context here, so please lmk if you disagree. Should this be returning an error rather than silently doing nothing? Or, is it possible that it could be intentional to set the BaseURL to an empty string?

jessepeterson replied:
API guard is done here: c77808e90a2965c6791a8de56a6cf61cea4c0b9b

zwass commented on storage/storagetest/storagetest.go:
```diff
@@ -0,0 +1,277 @@
+		tokens, err := storage.RetrieveAuthTokens(ctx, name)
+		// Currently just checking that it returns an error when the resource is not found.
+		//
+		// TODO(lucas): Define an "resource not found" error type as part of the interface
```
Good idea. Waiting on feedback from @jessepeterson to do that?

lucasmrod replied:
Oh I added it but wanted to check first if we all agree it's a good idea to define it and return it in all methods when the resource (config, tokens, pki) is not found.

zwass reviewed (commented):
Stoked to see this up!

jessepeterson commented on cmd/depsyncer/main.go:
```diff
@@ -31,7 +31,7 @@ func main() {
 		flADebug  = flag.Bool("debug-assigner", false, "additional debug logging of the device assigner")
 		flStorage = flag.String("storage", "", "storage backend")
-		flDSN     = flag.String("storage-dsn", "", "storage data source name")
+		flDSN     = flag.String("storage-dsn", "file", "storage data source name")
```
I think this meant to be on the `flStorage` flag?

lucasmrod replied:
Oops

jessepeterson reviewed (commented):
This is an amazing contribution! Thanks so much for this! Here's some quick thoughts and I'll try to get time soon, hopefully early next week for more. :)

Bunch of comments inline. :)

I have some reservations about having an opinionated docker-compose front and center. Let me think more on that (maybe it's as simple as placing the compose file in storage/storagetest/?).

The cmd/cli.go -> parse/storage.go is interesting. What was the thinking there?

Also - don't forget to update the docs (docs/operations-guide.md) for the new storage backend. :)

lucasmrod commented on storage/file/file.go:
```diff
@@ -119,7 +128,7 @@ func (s *FileStorage) RetrieveAssignerProfile(_ context.Context, name string) (s
 		var stat fs.FileInfo
 		stat, err = os.Stat(s.profileFilename(name))
-		if err != nil {
+		if err == nil {
```
Caught by tests! :)

lucasmrod commented on storage/file/file.go:
```diff
@@ -62,10 +65,14 @@ func (s *FileStorage) tokenpkiFilename(name, kind string) string {
-	return tokens, decodeJSONfile(s.tokensFilename(name), tokens)
+	err := decodeJSONfile(s.tokensFilename(name), tokens)
+	if err != nil {
+		return nil, err
```
A convention I've seen many projects use is to try (when possible) to always return zero-values when error is non-nil. In this case, the zero-value of a pointer is `nil`.

jessepeterson replied:
So I'm fine with this specific code change, although its a bit verbose. But in general I don't think its necessary to add a bunch of boilerplate to assure a zero-values for an error. Discussed here: https://macadmins.slack.com/archives/C07ME1284/p1663867171949419

lucasmrod commented on storage/mysql/mysql.go:
```diff
@@ -0,0 +1,294 @@
+	"errors"
+	"time"
+
+	_ "github.com/go-sql-driver/mysql"
```
Ah, good idea.

lucasmrod commented on storage/mysql/mysql.go:
```diff
@@ -0,0 +1,294 @@
+		&tokens.ConsumerSecret,
+		&tokens.AccessToken,
+		&tokens.AccessSecret,
+		&tokens.AccessTokenExpiry,
```
Oh, will check this. What was the failure?

lucasmrod commented on storage/file/file.go:
```diff
@@ -20,6 +21,8 @@ type FileStorage struct {
 	path string
 }
 
+var _ storage.AllStorage = (*FileStorage)(nil)
```
Convention that serves both as a "compile time check" and as a form of documentation for readers that `FileStorage` implements `storage.AllStorage` interface.

jessepeterson replied:
I think this creates an unnecessary coupling. How do you feel about removing it? As Go is statically typed an implementation must conform to it's use. `AllStorage` isn't particularly useful outside of `main.go` so I'd not like to couple these.

Let me know what you think. :)

lucasmrod commented on storage/mysql/mysql_test.go:
```diff
@@ -0,0 +1,86 @@
+
+	storagetest.Run(t, func(t *testing.T) storage.AllStorage {
+		dbName := initTestDB(t)
+		testDSN := fmt.Sprintf("nanodep:insecure@tcp(localhost:4242)/%s?charset=utf8mb4&parseTime=true&loc=UTC", dbName)
```
Ah makes sense. Will give this a thought.

lucasmrod commented on storage/mysql/mysql.go:
```diff
@@ -0,0 +1,294 @@
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			// an 'empty' config is valid
+			return &client.Config{}, nil
```
I wanted to mimick `FileStorage.RetrieveConfig`:
https://github.com/micromdm/nanodep/blob/077b73dfe75f5cf5ce12601ce1ca59bbb925bb0e/storage/file/file.go#L87-L98

jessepeterson replied:
Hmm, good call-out. I think the file-storage should be changed to match actually:

No config (missing) = nil
Present config (including an empty BaseURL) = returned.

They're treated the same in the config retriever interface implementation, as it happens. This should probably be documented there:

https://github.com/micromdm/nanodep/blob/a7538c5b9c87bca001a5bc636de747e1b1d10adc/client/client.go#L23-L25

lucasmrod commented on storage/mysql/mysql.go:
```diff
@@ -0,0 +1,294 @@
+// Returns an empty profile if it does not exist.
+func (s *MySQLStorage) RetrieveAssignerProfile(ctx context.Context, name string) (profileUUID string, modTime time.Time, err error) {
+	var (
+		profileUUID_ sql.NullString
```
Just wanted to make it clear that this nullable variable corresponds to the named return `profileUUID`.

lucasmrod commented on storage/mysql/mysql.go:
```diff
@@ -0,0 +1,294 @@
+		tokens.ConsumerSecret,
+		tokens.AccessToken,
+		tokens.AccessSecret,
+		tokens.AccessTokenExpiry,
```
As per tests it looks like it's working. (with the `parseTime` option now removed)

lucasmrod commented:
I believe I've address all the comments.

> I have some reservations about having an opinionated docker-compose front and center. Let me think more on that (maybe it's as simple as placing the compose file in storage/storagetest/?).

I've moved it to `storage/mysql/`. Let me know if that works.

> The cmd/cli.go -> parse/storage.go is interesting. What was the thinking there?

Originally I didn't want mysql imports in `storage/storage.go` (where we define the abstract interface), thus I moved the one parsing method to its own package `parse`.

> Also - don't forget to update the docs (docs/operations-guide.md) for the new storage backend. :)

Done.

Appreciate the feedback!

jessepeterson commented on parse/storage.go:
```diff
@@ -0,0 +1,28 @@
+package parse
```
I like the idea of moving this out of the `cmd/` tree. But can we rename this package to just `cli` to match https://github.com/micromdm/nanomdm/commit/ff8b7af3437525755d0d6b05d3c710590b454b6a?

jessepeterson reviewed (changes requested):
sorry for the unacceptably huge delay! just a few changes. would love your input on the general Go idiomatic/convention stuff — but it sounds like we might disagree on a few small things :). hopefully we can resolve!

lucasmrod commented:
> Hmm, good call-out. I think the file-storage should be changed to match actually:
No config (missing) = nil
Present config (including an empty BaseURL) = returned.

I've amended the two storage implementations to behave this way, and added a test for `DefaultConfigRetreiver`.

lucasmrod commented:
@jessepeterson One last change that can be added to this PR, or open a separate one if this one is merged-as is: https://github.com/fleetdm/nanodep/pull/1

jessepeterson commented:
@lucasmrod let's do a separate PR and keep this contained :)

also I think you need to 'request review' or whatever - it's showing as changes still requested :)

thanks again!

lucasmrod commented:
> let's do a separate PR and keep this contained :)

Cool. I'll create another PR once this one is merged to `main`.

jessepeterson commented on client/client.go:
```diff
@@ -21,7 +21,11 @@ type Config struct {
 
 type ConfigRetriever interface {
-	RetrieveConfig(context.Context, string) (*Config, error)
+	// RetrieveConfig reads the JSON DEP config of a DEP name.
```
❤️

jessepeterson reviewed (approved):
thanks much for this contribution @lucasmrod !
